File: Server/ReasnAPI/ReasnAPI/Models/API/EventResponse.cs
using ReasnAPI.Models.DTOs;
using ReasnAPI.Models.Enums;

namespace ReasnAPI.Models.API
{
    public class EventResponse(Participants participants)
    {
        public string Name { get; set; } = null!;
        public int AddressId { get; set; }
        public AddressDto AddressDto { get; set; } = null!;
        public string Description { get; set; } = null!;
        public Organizer Organizer { get; set; } = null!;
        public DateTime StartAt { get; set; }
        public DateTime EndAt { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        public string Slug { get; set; } = null!;
        public EventStatus Status { get; set; }
        public List<TagDto>? Tags { get; set; }
        public List<ParameterDto>? Parameters { get; set; }
        public Participants Participants { get; set; } = participants;
    }

}

File: Client/reasn-client/packages/common/src/schemas/EventResponse.ts
import { z } from "zod";
import { AddressDtoSchema } from "@reasn/common/src/schemas/AddressDto";
import ModelMappingError from "@reasn/common/src/errors/ModelMappingError";
import { TagDtoSchema } from "@reasn/common/src/schemas/TagDto";
import { ParameterDtoSchema } from "@reasn/common/src/schemas/ParameterDto";
import { EventStatus } from "@reasn/common/src/enums/schemasEnums";

export const EventRespsoneSchema = z.object({
  name: z.string().max(64),
  addressId: z.number(),
  address: AddressDtoSchema,
  description: z.string().max(4048),
  organizer: z.object({
    username: z.string(),
    image: z.string(),
  }),
  startAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  endAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  createdAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  updatedAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  slug: z
    .string()
    .max(128)
    .regex(/^[\p{L}\d]+[\p{L}\d-]*$/u)
    .nullable(),
  status: z.nativeEnum(EventStatus),
  tags: z.array(TagDtoSchema),
  parameters: z.array(ParameterDtoSchema),
  participants: z.object({
    interested: z.number(),
    participating: z.number(),
  }),
});

export type EventResponse = z.infer<typeof EventRespsoneSchema>;

export const EventResponseMapper = {
  fromObject: (entity: object): EventResponse => {
    const result = EventRespsoneSchema.safeParse(entity);
    if (!result.success) {
      throw new ModelMappingError(
        "EventResponse",
        result.error.message,
        result.error,
      );
    }
    return result.data;
  },
  fromJSON: (jsonEntity: string): any => {
    if (!jsonEntity) {
      throw new ModelMappingError("EventResponse", "Empty JSON string");
    }
    const result = EventRespsoneSchema.safeParse(JSON.parse(jsonEntity));
    if (!result.success) {
      throw new ModelMappingError(
        "EventResponse",
        result.error.message,
        result.error,
      );
    }
    return result.data;
  },
};
